storagemarket/deal_handler.go

```diff
@@ -37,3 +37,35 @@ func (d *dealHandler) subscribeUpdates() (event.Subscription, error) {
 	}
 	return sub, nil
 }
+
+func (dh *dealHandler) cancel() error {
+	dh.transferMu.Lock()
+	defer dh.transferMu.Unlock()
+
+	if dh.transferFinished {
+		return dh.transferErr
+	}
+
+	dh.transferCancel()
+
+	select {
+	case err := <-dh.transferDone:
+		dh.transferFinished = true
+		dh.transferErr = err
+		return err
```

Comment:
I don't think this will work if cancel is called twice before the transfer completes.
It's a little edge-casey but worth thinking about.

Reply:
@dirkmc The locking on line 42-43 should take care of that.

Reply:
Ah yep makes sense 👍